Eta on NixOS

Pinging @puffnfresh :)

I am following this:
https://eta-lang.org/docs/user-guides/eta-user-guide/installation/nix

```bash
$ git clone --branch stable --recursive --depth 1 https://github.com/typelead/eta
$ cd eta
$ nix-shell -A eta-build-shell
```

The `nix-shell` command fails with the following:

```bash
Linking Setup ...
[ 1 of 84] Compiling Distribution.Simple.Program.Internal ( Distribution/Simple/Program/Internal.hs, /build/Distribution/Simple/Program/Internal.o )
[ 2 of 84] Compiling Distribution.Compat.MonadFail ( Distribution/Compat/MonadFail.hs, /build/Distribution/Compat/MonadFail.o )
[ 3 of 84] Compiling Distribution.Compat.GetShortPathName ( Distribution/Compat/GetShortPathName.hs, /build/Distribution/Compat/GetShortPathName.o )
[ 4 of 84] Compiling Distribution.GetOpt ( Distribution/GetOpt.hs, /build/Distribution/GetOpt.o )
[ 5 of 84] Compiling Distribution.Compat.Binary ( Distribution/Compat/Binary.hs, /build/Distribution/Compat/Binary.o )
[ 6 of 84] Compiling Distribution.Compat.Semigroup ( Distribution/Compat/Semigroup.hs, /build/Distribution/Compat/Semigroup.o )
[ 7 of 84] Compiling Distribution.Lex ( Distribution/Lex.hs, /build/Distribution/Lex.o )
[ 8 of 84] Compiling Distribution.TestSuite ( Distribution/TestSuite.hs, /build/Distribution/TestSuite.o )
[ 9 of 84] Compiling Distribution.Simple.PreProcess.Unlit ( Distribution/Simple/PreProcess/Unlit.hs, /build/Distribution/Simple/PreProcess/Unlit.o )
[10 of 84] Compiling Distribution.Simple.CCompiler ( Distribution/Simple/CCompiler.hs, /build/Distribution/Simple/CCompiler.o )
[11 of 84] Compiling Distribution.PackageDescription.Utils ( Distribution/PackageDescription/Utils.hs, /build/Distribution/PackageDescription/Utils.o )
[12 of 84] Compiling Distribution.Compat.ReadP ( Distribution/Compat/ReadP.hs, /build/Distribution/Compat/ReadP.o )
[13 of 84] Compiling Distribution.ReadE ( Distribution/ReadE.hs, /build/Distribution/ReadE.o )
[14 of 84] Compiling Distribution.Verbosity ( Distribution/Verbosity.hs, /build/Distribution/Verbosity.o )
[15 of 84] Compiling Distribution.Text ( Distribution/Text.hs, /build/Distribution/Text.o )
[16 of 84] Compiling Distribution.ModuleName ( Distribution/ModuleName.hs, /build/Distribution/ModuleName.o )
[17 of 84] Compiling Distribution.System ( Distribution/System.hs, /build/Distribution/System.o )
[18 of 84] Compiling Distribution.Version ( Distribution/Version.hs, /build/Distribution/Version.o )

Distribution/Version.hs:132:10:
    No instance for (Binary Version)
      arising from a use of ‘binary-0.7.5.0:Data.Binary.Class.$gdmput’
    In the expression: binary-0.7.5.0:Data.Binary.Class.$gdmput
    In an equation for ‘put’:
        put = binary-0.7.5.0:Data.Binary.Class.$gdmput
    In the instance declaration for ‘Binary VersionRange’
[22 of 84] Compiling Language.Haskell.Extension ( Language/Haskell/Extension.hs, /build/Language/Haskell/Extension.o )
[26 of 84] Compiling Distribution.Compat.Exception ( Distribution/Compat/Exception.hs, /build/Distribution/Compat/Exception.o )
[27 of 84] Compiling Distribution.Compat.Internal.TempFile ( Distribution/Compat/Internal/TempFile.hs, /build/Distribution/Compat/Internal/TempFile.o )
[28 of 84] Compiling Distribution.Compat.CopyFile ( Distribution/Compat/CopyFile.hs, /build/Distribution/Compat/CopyFile.o )
[41 of 84] Compiling Distribution.Compat.Environment ( Distribution/Compat/Environment.hs, /build/Distribution/Compat/Environment.o )
[72 of 84] Compiling Distribution.Compat.CreatePipe ( Distribution/Compat/CreatePipe.hs, /build/Distribution/Compat/CreatePipe.o )
builder for '/nix/store/al803zvbmbxshc61mg608v5yg80fwgyr-Cabal-1.24.2.0.drv' failed with exit code 1
cannot build derivation '/nix/store/gvldir7pcfs473578q34xllsfyh38y56-etlas-1.3.0.0.drv': 1 dependencies couldn't be built
error: build of '/nix/store/5479hwfpa9y53643vx7b12i4i604lpvz-eta-0.7.2.1.drv', '/nix/store/g5xyb6y6116fsw5wmspjd66lsxg68nr4-eta-build-0.0.0.drv', '/nix/store/gvldir7pcfs473578q34xllsfyh38y56-etlas-1.3.0.0.drv', '/nix/store/pxvhqxjg2p490c4ynnslpq2x62p9dxd6-eta-pkg-0.7.2.1.drv' failed
```

Note that I am using NixOS unstable, not quite sure that would really make a difference:

```bash
λ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`                               
 - host os: `Linux 4.17.10, NixOS, 18.09.git.443cbd7 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.0.4`
 - channels(root): `"nixos-18.09pre145679.dae9cf6106d"`
 - channels(nequi): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

So I figured I'd try a Gradle build:

https://eta-lang.org/docs/user-guides/eta-user-guide/installation/gradle

I checked out `eta-init` and ran the following:

```bash
λ nix-shell -p openjdk8

[nix-shell:~/dev/eta/eta-init]$ ./gradlew run
Starting a Gradle Daemon (subsequent builds will be faster)
> Task :setupEnvironmentEta FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':setupEnvironmentEta'.
> IOException occurred when executing command [/home/nequi/.gradle/caches/etlas/1.4.0.0/etlas, init] with workingDir null

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 8s
1 actionable task: 1 executed
```

Here is the stack trace:

<details>
<summary>Expand stacktrace</summary>

<pre>
[nix-shell:~/dev/eta/eta-init]$ ./gradlew run --stacktrace
> Task :setupEnvironmentEta FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':setupEnvironmentEta'.
> IOException occurred when executing command [/home/nequi/.gradle/caches/etlas/1.4.0.0/etlas, init] with workingDir null

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':setupEnvironmentEta'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:103)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:73)
	at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
	at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:256)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:249)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:238)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:663)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:596)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.api.GradleException: IOException occurred when executing command [/home/nequi/.gradle/caches/etlas/1.4.0.0/etlas, init] with workingDir null
	at com.typelead.gradle.utils.CommandLine.start(CommandLine.java:207)
	at com.typelead.gradle.utils.CommandLine.executeAndConsumeOutputWithProcess(CommandLine.java:143)
	at com.typelead.gradle.utils.EtlasCommand.getWelcomeMessage(EtlasCommand.java:92)
	at com.typelead.gradle.eta.tasks.EtaSetupEnvironment.ensureTelemetryPreferencesAndUpdate(EtaSetupEnvironment.java:259)
	at com.typelead.gradle.eta.tasks.EtaSetupEnvironment.setupEnvironment(EtaSetupEnvironment.java:125)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:794)
	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:761)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:124)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:113)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:95)
	... 29 more
Caused by: java.io.IOException: Cannot run program "/home/nequi/.gradle/caches/etlas/1.4.0.0/etlas": error=2, No such file or directory
	at com.typelead.gradle.utils.CommandLine.start(CommandLine.java:205)
	... 47 more
Caused by: java.io.IOException: error=2, No such file or directory
	... 48 more


* Get more help at https://help.gradle.org

BUILD FAILED in 0s
1 actionable task: 1 executed
</pre>
</details>


The `etlas` binary does exist but not all libraries can be linked:

```bash
[nix-shell:~/dev/eta/eta-init]$ ls -la /home/nequi/.gradle/caches/etlas/1.4.0.0/etlas
-rwxr--r-- 1 nequi users 26834736 Jul 26 16:39 /home/nequi/.gradle/caches/etlas/1.4.0.0/etlas

[nix-shell:~/dev/eta/eta-init]$ ldd !$
ldd /home/nequi/.gradle/caches/etlas/1.4.0.0/etlas
	linux-vdso.so.1 (0x00007ffe231dc000)
	libz.so.1 => not found
	librt.so.1 => /nix/store/83lrbvbmxrgv7iz49mgd42yvhi473xp6-glibc-2.27/lib/librt.so.1 (0x00007d6578256000)
	libutil.so.1 => /nix/store/83lrbvbmxrgv7iz49mgd42yvhi473xp6-glibc-2.27/lib/libutil.so.1 (0x00007d6578053000)
	libdl.so.2 => /nix/store/83lrbvbmxrgv7iz49mgd42yvhi473xp6-glibc-2.27/lib/libdl.so.2 (0x00007d6577e4f000)
	libgmp.so.10 => not found
	libm.so.6 => /nix/store/83lrbvbmxrgv7iz49mgd42yvhi473xp6-glibc-2.27/lib/libm.so.6 (0x00007d6577aba000)
	libpthread.so.0 => /nix/store/83lrbvbmxrgv7iz49mgd42yvhi473xp6-glibc-2.27/lib/libpthread.so.0 (0x00007d657789b000)
	libgcc_s.so.1 => /nix/store/83lrbvbmxrgv7iz49mgd42yvhi473xp6-glibc-2.27/lib/libgcc_s.so.1 (0x00007d6577685000)
	libc.so.6 => /nix/store/83lrbvbmxrgv7iz49mgd42yvhi473xp6-glibc-2.27/lib/libc.so.6 (0x00007d65772d1000)
	/lib64/ld-linux-x86-64.so.2 => /nix/store/83lrbvbmxrgv7iz49mgd42yvhi473xp6-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007d657845e000)
```

--- a/default.nix
+++ b/default.nix
@@ -17,14 +17,6 @@ let
       src = rewriteRelative eta-nix args.src;
     });
 
-    etlas = haskell.lib.overrideCabal super.etlas (drv: {
-      # Nix should only compile Setup.hs with setup-depends, but it doesn't:
-      # https://github.com/NixOS/nixpkgs/issues/24809
-      preCompileBuildDriver = ''
-        ${drv.preCompileBuildDriver or ""}
-        setupCompileFlags+=" -hide-package=etlas-cabal"
-      '';
-    });
     eta = haskell.lib.overrideCabal super.eta (drv: {
       # Makes the build a bit faster
       src = onlyFiles ["compiler" "include" "eta" "eta.cabal" "LICENSE" "tests"] drv.src;

